include/seqan3/search/algorithm/detail/search.hpp

```diff
@@ -0,0 +1,280 @@
+    std::vector<typename index_t::iterator_type> internal_hits;
+    auto internal_delegate = [&internal_hits, &max_error](auto const & it/*, detail::search_params const error_left*/)
+    {
+        internal_hits.push_back(it);
+        // internal_hits.push_back({it, max_error - error_left});
+    };
+
+    // choose strategy
+    if constexpr (contains<id::strategy_best>(cfg))
+    {
+        detail::search_params max_error2{max_error};
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<true>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+    }
+    else if constexpr (contains<id::strategy_all_best>(cfg))
+    {
+        detail::search_params max_error2{max_error};
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<false>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+    }
+    else if constexpr (contains<id::strategy_strata>(cfg))
+    {
+        detail::search_params max_error2{max_error};
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<true>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+        if (!internal_hits.empty())
+        {
+            internal_hits.clear(); // don't clear when using Optimum Search Schemes with lower error bounds
+            uint8_t const s = seqan3::get<id::strategy_strata>(cfg);
+            max_error2.total += s - 1;
+            detail::search_trivial<false>(index, query, max_error2, internal_delegate);
+        }
+    }
+    else // "strategy_all" or not specified
+    {
+        detail::search_trivial<false>(index, query, max_error, internal_delegate);
+    }
+
+    // TODO: filter hits and only do it when necessary (depending on error types)
+    // _filter_hits(hits);
+
+    // output iterators or text_positions
+    if constexpr (contains<id::output_index_iterator>(cfg))
+    {
+        return internal_hits;
+    }
+    else
+    {
+        std::vector<typename index_t::size_type> hits;
+        // std::vector<std::tuple<typename index_t::size_type, detail::search_params>> hits;
+        // for (auto const & [it, error] : internal_hits)
+        if constexpr (contains<id::strategy_best>(cfg))
+        {
+            // only one iterator is reported but it might contain more than one text position
+            if (!internal_hits.empty())
+            {
+                auto const & text_pos = internal_hits[0].lazy_locate();
+                hits.push_back(text_pos[0]);
+            }
+        }
+        else
+        {
+            for (auto const & it : internal_hits)
```

Comment:
that is now the confusing thing.
Usually i would expect  an iterator to be in O(1) copyable, so I don't pass it by ref but by value. But this is the index iterator with a state?

Reply:
Yes, I'm not sure about this. I could copy it in O(1). But it would still be up to 100 bytes vs. 8 byte for a reference.